docs_src/design/adr/security/0017-consul-security.md

```diff
@@ -0,0 +1,111 @@
+# Securing access to Consul
+
+## Status
+
+** proposed **
+
+## Context
+
+This ADR defines the motiviation and approach used to secure access
+to the Consul component in the EdgeX archticture
+for *security-enabled configurations only*.
+(Non-secure configuations continue to use Consul in
+anonymous read-write mode.)
+
+Consul provides several services for the EdgeX architecture:
+
+- Service registry (see ADR in references below)
+- Service health monitoring
+- Mutable configuration data
+
+Use of the services provided by Consul is optional on a service-by-service basis.
+Use of the registry is controlled by the `-r` or `--registry` flag provided to an EdgeX service.
+Use of mutable configuration data is controlled by the `-cp` or `--configProvider` flag provided to an EdgeX service.
+When Consul is enabled as a configuration provider,
+the `configuration.toml` is parsed into individual settings
+and seeded into the Consul key-value store on the first start of a service.
+Configuration reads and writes are then done to Consul if it is specified as the configuration provider,
+otherwise the static `configuration.toml` is used.
+Writes to the `[Writable]` section in Consul trigger per-service callbacks
+notifying the application of the changed data.
+
+Since configuration data can affect the runtime behavior of services,
+compensating controls must be introduced in order to mitigate the risks introduced
+by moving configuration from a static file into to an HTTP-accessible service with mutable state.
+
+The current practice is that Consul is exposed via unencrypted HTTP in anonymous read/write mode
+to all processes and EdgeX services running on the host machine.
+
+## Decision
+
+Consul will be configured with access control list (ACL) functionality enabled,
+and each EdgeX service will utilize a Consul access token to authenticate to Consul.
+Consul access tokens will be requested from the Vault Consul secrets engine
+(to avoid introducing additional bootstrapping secrets).
+
+DNS will be disabled via configuration as it is not used in EdgeX.
+
+**Consul Access Via API Gateway**
+
+Consul can be configured to be exposed via the EdgeX API Gateway
+in order to provide remote configuration capabilities.
+The UI can also be rebased (in theory) via the `ui_content_path`
+[setting](https://www.consul.io/docs/agent/options#ui_config_enabled)
+in order to be compatible with the API Gateway pathing conventions.
+
+Consul supports authentication via the `X-Consul-Token` HTTP header,
+the `Authorization: Bearer` header,
+or via a `?token=` query parameter (insecure).
+The API gateway should be configured to simply pass HTTP requests
+and rely on Consul's authentication mechanisms for both API and UI accesses.
```

Comment:
With the API gateway, once a user has been defined for JWT Auth, the tokens can be generated on an external system. How does this work for Consul?

Reply:
The proposal calls for generating an extra management toked for remote management purposes.  This token would be able to read/write the entire KV store and incidentally to generate extra tokens if needed.